### ISSUE TITLE
Port from bespoke assertions of snapbox

### ISSUE THREAD

**Issue by epage** [rust-lang/cargo#14039]
Traditionally, [`cargo-test-support`](https://doc.rust-lang.org/nightly/nightly-rustc/cargo_test_support/index.html) has had bespoke assertions, requiring hand implemented diff algorithms.  As there is less of a community around this, the knowledge is more specialized, its less likely to be documented, and we are on our own for feature development.

In #13980 and #14031, we introduced the use of [snapbox](https://crates.io/crates/snapbox) as replacement for our own assertions and need to work to migrate to them.

Aside: doc updates related to this transition
- #14243
- #14266
- #14269
- #14268
- #14270
- #14272

### 1. Select a file to port

The files should have `#![allow(deprecated)]` at the top, e.g.
```console
$ git pull --rebase
$ rg '#!.allow.deprecated' tests
```

Please check the comments on this issue for anyone to have claimed the file and then post that you claim the file

### 2. Remove `#![allow(deprecated)]` to identify what work is needed

If nothing, congrats, that was easy!

### 3. Resolve basic deprecations

Replace `Execs::with_stdout("...")` with
```rust
use cargo_test_support::prelude::*;
use cargo_test_support::str;

.with_stdout_data(str![])
```

Replace `Execs::with_stderr("...")` with
```rust
use cargo_test_support::prelude::*;
use cargo_test_support::str;

.with_stderr_data(str![])
```
(prelude is only needed for some steps)

> *Side note:* If the test is specifically wanting to assert an empty string, use `""` rather than `str![]` so that people are less likely to accidentally do a snapshot update to bless output that gets added.

Commit these changes

Run `SNAPSHOTS=overwrite cargo test && cargo check --test testsuite`.  In rare cases, `SNAPSHOTS=overwrite` may make bad edits.  Feel free to create a reproduction case and [create an issue](https://github.com/assert-rs/snapbox/issues/new/choose).  They are usually obvious how to fix, so don't worry.

Diff the snapshots.  Is there anything machine or run specific in them that previously was elided out with `[..]`?
- Can you auto-redact it like 5ea1c8f?  If so, revert the snapshot updates, add the auto-redaction (as a separate commit before any test changes), and re-run the snapshots
- Otherwise, replace it with `[..]`

Once its working, amend your commit with the snapshots

### 4. Resolve non-literal deprecations

Like Step 3, but not just `with_stdout("...")` but with variables or `format!`.

Evaluate whether a literal could be used
- Is the variable important for showing that output remains unchanged between instances, then keep it
- Are we specifically testing for what we are using `format!` for, then keep it
  - If not, see if it can be expressed with an auto-redaction

So this can end up with either
- `with_std*_data(expression)`
- `with_std*_data(str![])` like in Step 3

### 5. Repeat with "straight forward" deprecations

- `with_stdout_unordered(expected)` -> `with_stdout_data(expected.unordered())`
- `with_stderr_unordered(expected)` -> `with_stderr_data(expected.unordered())`
- `with_json(expected)` -> `with_stdout_data(expected.json_lines())` or `with_stdout_data(expected.json())`
  - May also require `.unordered()`

### 6. Contains / Does not Contains deprecations

A judgement needs to be made for how to handle these.

A `contains` can be modeled by surrounding a entry with `...` (text) or `"...",` (json lines).  The question is whether we should still do this or use a different method.
- If replacing a `contains` with an equality check, use multi-line globs (`...`) for rustc errors that weren't previously matched to minimize tying Cargo's tests to the exact output of rustc
- Multiple `contains` can be merged into a single `eq` with `..` interspersed (or one `...` and `.unordered()`).  For example, see 3054936cab3d4627d5d1d878426c4fd88fab41f8
  - When using `unordered`, likely string literals, rather than `str![]` should be used, as these can't correctly be updated

A `does_not_contain`/`line_without` cannot be modeled at this time.   The challenge with these is that they are brittle and you can't tell when they are no longer testing for what you think because the output changed.  We should evaluate what to do with these on a case-by-case basis.

When in doubt, feel free to put `#[allow(deprecated)]` on a statement and move on.  We can come back to these later.

**Comment by weihanglo** [comment]
## Porting progress

- [ ] tests/testsuite/advanced_env.rs
- [x] tests/testsuite/alt_registry.rs #14031
- [ ] tests/testsuite/artifact_dep.rs
- [ ] tests/testsuite/artifact_dir.rs
- [ ] tests/testsuite/bad_config.rs
- [ ] tests/testsuite/bad_manifest_path.rs
- [ ] tests/testsuite/bench.rs
- [ ] tests/testsuite/binary_name.rs
- [ ] tests/testsuite/build.rs
- [ ] tests/testsuite/build_plan.rs
- [ ] tests/testsuite/build_script.rs
- [ ] tests/testsuite/build_script_env.rs
- [ ] tests/testsuite/build_script_extra_link_arg.rs
- [ ] tests/testsuite/cache_lock.rs
- [ ] tests/testsuite/cache_messages.rs
- [ ] tests/testsuite/cargo_alias_config.rs
- [ ] tests/testsuite/cargo_command.rs
- [ ] tests/testsuite/cargo_env_config.rs
- [ ] tests/testsuite/cargo_features.rs
- [ ] tests/testsuite/cargo_targets.rs
- [ ] tests/testsuite/cfg.rs
- [ ] tests/testsuite/check.rs
- [ ] tests/testsuite/check_cfg.rs
- [ ] tests/testsuite/clean.rs
- [ ] tests/testsuite/collisions.rs
- [ ] tests/testsuite/concurrent.rs
- [ ] tests/testsuite/config.rs
- [ ] tests/testsuite/config_cli.rs
- [ ] tests/testsuite/config_include.rs
- [ ] tests/testsuite/corrupt_git.rs
- [ ] tests/testsuite/credential_process.rs
- [ ] tests/testsuite/cross_compile.rs
- [ ] tests/testsuite/cross_publish.rs
- [ ] tests/testsuite/custom_target.rs
- [ ] tests/testsuite/death.rs
- [ ] tests/testsuite/dep_info.rs
- [ ] tests/testsuite/diagnostics.rs
- [ ] tests/testsuite/direct_minimal_versions.rs
- [ ] tests/testsuite/directory.rs
- [ ] tests/testsuite/doc.rs
- [ ] tests/testsuite/docscrape.rs
- [ ] tests/testsuite/edition.rs
- [ ] tests/testsuite/error.rs
- [ ] tests/testsuite/features.rs
- [ ] tests/testsuite/features2.rs
- [ ] tests/testsuite/features_namespaced.rs
- [ ] tests/testsuite/fetch.rs
- [ ] tests/testsuite/fix.rs
- [ ] tests/testsuite/fix_n_times.rs
- [ ] tests/testsuite/freshness.rs
- [ ] tests/testsuite/future_incompat_report.rs
- [ ] tests/testsuite/generate_lockfile.rs
- [ ] tests/testsuite/git.rs
- [ ] tests/testsuite/git_auth.rs
- [ ] tests/testsuite/git_gc.rs
- [ ] tests/testsuite/git_shallow.rs
- [ ] tests/testsuite/glob_targets.rs
- [ ] tests/testsuite/global_cache_tracker.rs
- [ ] tests/testsuite/help.rs
- [ ] tests/testsuite/https.rs
- [ ] tests/testsuite/inheritable_workspace_fields.rs
- [ ] tests/testsuite/install.rs
- [ ] tests/testsuite/install_upgrade.rs
- [ ] tests/testsuite/jobserver.rs
- [ ] tests/testsuite/lints_table.rs
- [ ] tests/testsuite/list_availables.rs
- [ ] tests/testsuite/local_registry.rs
- [ ] tests/testsuite/locate_project.rs
- [ ] tests/testsuite/lockfile_compat.rs
- [ ] tests/testsuite/login.rs
- [ ] tests/testsuite/logout.rs
- [ ] tests/testsuite/lto.rs
- [ ] tests/testsuite/main.rs
- [ ] tests/testsuite/member_discovery.rs
- [ ] tests/testsuite/member_errors.rs
- [ ] tests/testsuite/message_format.rs
- [ ] tests/testsuite/messages.rs
- [ ] tests/testsuite/metabuild.rs
- [ ] tests/testsuite/metadata.rs
- [ ] tests/testsuite/minimal_versions.rs
- [ ] tests/testsuite/multitarget.rs
- [ ] tests/testsuite/net_config.rs
- [ ] tests/testsuite/new.rs
- [ ] tests/testsuite/offline.rs
- [ ] tests/testsuite/old_cargos.rs
- [ ] tests/testsuite/open_namespaces.rs
- [ ] tests/testsuite/owner.rs
- [ ] tests/testsuite/package.rs
- [ ] tests/testsuite/package_features.rs
- [ ] tests/testsuite/patch.rs
- [ ] tests/testsuite/path.rs
- [ ] tests/testsuite/paths.rs
- [ ] tests/testsuite/pkgid.rs
- [ ] tests/testsuite/precise_pre_release.rs
- [ ] tests/testsuite/proc_macro.rs
- [ ] tests/testsuite/profile_config.rs
- [ ] tests/testsuite/profile_custom.rs
- [ ] tests/testsuite/profile_overrides.rs
- [ ] tests/testsuite/profile_targets.rs
- [ ] tests/testsuite/profile_trim_paths.rs
- [ ] tests/testsuite/profiles.rs
- [ ] tests/testsuite/progress.rs
- [ ] tests/testsuite/pub_priv.rs
- [ ] tests/testsuite/publish.rs
- [ ] tests/testsuite/publish_lockfile.rs
- [ ] tests/testsuite/read_manifest.rs
- [ ] tests/testsuite/registry.rs
- [ ] tests/testsuite/registry_auth.rs
- [ ] tests/testsuite/rename_deps.rs
- [ ] tests/testsuite/replace.rs
- [ ] tests/testsuite/required_features.rs
- [ ] tests/testsuite/run.rs
- [ ] tests/testsuite/rust_version.rs
- [ ] tests/testsuite/rustc.rs
- [ ] tests/testsuite/rustc_info_cache.rs
- [ ] tests/testsuite/rustdoc.rs
- [ ] tests/testsuite/rustdoc_extern_html.rs
- [ ] tests/testsuite/rustdocflags.rs
- [ ] tests/testsuite/rustflags.rs
- [ ] tests/testsuite/rustup.rs
- [ ] tests/testsuite/script.rs
- [ ] tests/testsuite/search.rs
- [ ] tests/testsuite/shell_quoting.rs
- [ ] tests/testsuite/source_replacement.rs
- [ ] tests/testsuite/ssh.rs
- [ ] tests/testsuite/standard_lib.rs
- [ ] tests/testsuite/test.rs
- [ ] tests/testsuite/timings.rs
- [ ] tests/testsuite/tool_paths.rs
- [ ] tests/testsuite/tree.rs
- [ ] tests/testsuite/tree_graph_features.rs
- [ ] tests/testsuite/unit_graph.rs
- [ ] tests/testsuite/update.rs
- [ ] tests/testsuite/vendor.rs
- [ ] tests/testsuite/verify_project.rs
- [ ] tests/testsuite/version.rs
- [ ] tests/testsuite/warn_on_failure.rs
- [ ] tests/testsuite/weak_dep_features.rs
- [ ] tests/testsuite/workspaces.rs
- [ ] tests/testsuite/yank.rs

**Comment by heisen-li** [comment]
I would like to try the following file:
``` 
 tests/testsuite/build_script_env.rs
 tests/testsuite/build_script_extra_link_arg.rs
 tests/testsuite/cache_lock.rs
 tests/testsuite/cache_messages.rs
 tests/testsuite/cargo_alias_config.rs
 tests/testsuite/cargo_command.rs
 tests/testsuite/cargo_config/mod.rs
 tests/testsuite/cargo_env_config.rs
 tests/testsuite/cargo_features.rs
 tests/testsuite/cargo_targets.rs
```

**Comment by epage** [comment]
#14041 highlights a blocker for `tests/testsuite/message_format.rs` (and other jsonlines tests): `snapbox` can only render the output as jsonlines and not our "pretty printed" variant which can make the messages harder to read.

**Comment by henry40408** [comment]
@weihanglo I'd like to claim `tests/testsuite/shell_quoting.rs`.

**Comment by choznerol** [comment]
@weihanglo I'd like to claim `tests/testsuite/help.rs`.

**Comment by choznerol** [comment]
@weihanglo I would like to claim `tests/testsuite/check_cfg.rs` and `tests/testsuite/freshness.rs` as well to prevent conflict with the fix for #14076.

**Comment by dieterplex** [comment]
@weihanglo going to pick up these:

```
tests/testsuite/collisions.rs
tests/testsuite/concurrent.rs
tests/testsuite/config.rs
tests/testsuite/config_cli.rs
tests/testsuite/config_include.rs
tests/testsuite/corrupt_git.rs
```

**Comment by henry40408** [comment]
@weihanglo I'd like to claim `tests/testsuite/minimal_versions.rs`

**Comment by henry40408** [comment]
@weihanglo I'd like to report that `with_stdout` or `with_stderr` doesn't appear in `tests/testsuite/member_discovery.rs`, so the file should be removed from the list.

By the way, I'd like to claim `tests/testsuite/timings.rs`.

**Comment by weihanglo** [comment]
@henry40408 would you mind removing `#[allow(deprecated)]` from memeber_discovery.rs?

**Comment by henry40408** [comment]
@weihanglo I'll remove it in https://github.com/rust-lang/cargo/pull/14082

**Comment by dieterplex** [comment]
woring on clean at #14096

**Comment by linyihai** [comment]
I'd like to claim `tests/testsuite/pub_priv.rs` since i  have added the test before

**Comment by dieterplex** [comment]
Claim these
- tests/testsuite/git.rs
- tests/testsuite/git_auth.rs

**Comment by linyihai** [comment]
claim `tests/testsuite/lints/unused_optional_dependencies.rs`

Since I am now modifying this test case, I can migrate it together :)

**Comment by dieterplex** [comment]
Cooking `tests/testsuite/lto.rs`

**Comment by epage** [comment]
I've created assert-rs/snapbox#348 to track what we need from snapbox for jsonlines.  See that issue for my proposed solution.

**Comment by epage** [comment]
I fixed assert-rs/snapbox#348  but found assert-rs/snapbox#351  and assert-rs/snapbox#352

**Comment by epage** [comment]
Looks like we've finished every file-level migration.  We have 124 remaining `#[allow(deprecated)]`s left.
